### PR TITLE
IKASAN-1581 adding flow and module runtime meta data functionality to…

### PR DESCRIPTION
**Pull request by mick-stewart73**
… the topology module and exposing the functionality via rest.

### CODE REVIEW

File: ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonFlowMetaDataProvider.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    @Override
+    public FlowMetaData deserialiseFlow(String metaData)
+    {
+        FlowMetaDataImpl result;
+
+        try
+        {
+            ObjectMapper mapper = new ObjectMapper();
```
Comment (andrzej-majewski) [inline ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonFlowMetaDataProvider.java]: the mapper should be initialised in constructor once. there is no need for creating new mapper every time.

File: ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonFlowMetaDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+    @Override
+    public String describeFlow(Flow flow)
+    {
+        String result;
+
+        try
+        {
+            FlowConfiguration configuration = flow.getFlowConfiguration();
+            flow.getName();
```
Comment (andrzej-majewski) [inline ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonFlowMetaDataProvider.java]: no need to call flow.getName() without assigning it to a value.

File: ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonFlowMetaDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            FlowMetaDataImpl flowMetaData = new FlowMetaDataImpl();
+            flowMetaData.setName(flow.getName());
+            flowMetaData.setConsumer(describeFlowElement(flowMetaData, configuration.getConsumerFlowElement()));
+
+            if(flow instanceof ConfiguredResource)
+            {
+                flowMetaData.setConfigurationId(((ConfiguredResource) flow).getConfiguredResourceId());
+            }
+
+            ObjectMapper mapper = new ObjectMapper();
```
Comment (andrzej-majewski) [inline ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonFlowMetaDataProvider.java]: Mapper should be initialised in constructor.

File: ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonModuleMetaDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            moduleMetaData.setDescription(module.getDescription());
+            moduleMetaData.setVersion(module.getVersion());
+
+            for(Flow flow: module.getFlows())
+            {
+                moduleMetaData.getFlows().add(flowMetaDataProvider
+                    .deserialiseFlow(flowMetaDataProvider.describeFlow(flow)));
+            }
+
+            ObjectMapper mapper = new ObjectMapper();
```
Comment (andrzej-majewski) [inline ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonModuleMetaDataProvider.java]: Mapper should be created in constructor

File: ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonModuleMetaDataProvider.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Override
+    public ModuleMetaData deserialiseModule(String module)
+    {
+        ModuleMetaDataImpl result;
+
+        try
+        {
+            ObjectMapper mapper = new ObjectMapper();
```
Comment (andrzej-majewski) [inline ikasaneip/topology/src/main/java/org/ikasan/topology/metadata/JsonModuleMetaDataProvider.java]: Same here mapper can be initialised in constructor

File: ikasaneip/topology/src/test/java/org/ikasan/topology/metadata/JsonFlowMetaDataProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    public static final String MULTI_RECIPIENT_FLOW_RESULT_JSON = "/data/multiRecipientFlow.json";
+    public static final String SINGLE_RECIPIENT_FLOW_RESULT_JSON = "/data/singleRecipientFlow.json";
+
+    @Test
+    public void test_simple_flow() throws IOException
+    {
+        JsonFlowMetaDataProvider jsonFlowMetaDataProvider = new JsonFlowMetaDataProvider();
+        String json = jsonFlowMetaDataProvider.describeFlow(createSimpleFlow());
+
+        Assert.assertEquals("JSON Result must equal!", loadDataFile(SIMPLE_FLOW_RESULT_JSON).replaceAll("\r", "").replaceAll("\n", "")
```
Comment (andrzej-majewski) [inline ikasaneip/topology/src/test/java/org/ikasan/topology/metadata/JsonFlowMetaDataProviderTest.java]: two options in here,
1) use lib which will help with compering and asserting JSON,  like https://www.baeldung.com/jsonassert
2) dont use a pretty print when you serialising object to JSON string., or allow injection of boolean flag which you can utilise in a test

Comment (mick-stewart73) [reply ikasaneip/topology/src/test/java/org/ikasan/topology/metadata/JsonFlowMetaDataProviderTest.java]: @andrzej-majewski Take a look now.

**Pull request merged**